diff --git a/src/base/functional/bind-internal.h b/src/base/functional/bind-internal.h
index 160664af5dd..4b46a2680b9 100644
--- a/src/base/functional/bind-internal.h
+++ b/src/base/functional/bind-internal.h
@@ -28,13 +28,24 @@
 //                       function signature adapters are applied.
 
 namespace v8::base::internal {
+
+template <typename T>
+struct is_std_function : std::false_type {};
+
+template <typename Signature>
+struct is_std_function<std::function<Signature>> : std::true_type {};
+
+template <typename T>
+inline constexpr bool is_std_function_v = is_std_function<std::remove_cvref_t<T>>::value;

 // True if `T` is completely defined.
 template <typename T>
 concept IsComplete = requires { sizeof(T); } ||
                      // Function types must be included explicitly since you
                      // cannot apply `sizeof()` to a function type.
-                     std::is_function_v<std::remove_cvref_t<T>>;
+                     std::is_function_v<std::remove_cvref_t<T>>||
+                     // MSVC STL compatibility fallback for std::function instances
+                     is_std_function_v<T>;

 // Packs a list of types to hold them in a single type.
 template <typename... Types>
@@ -86,6 +97,14 @@ concept HasOverloadedCallOp = requires {
 template <typename Signature>
 struct ExtractArgsImpl;

+// 2. Missing specialization to extract arguments from std::function
+template <typename R, typename... Args>
+struct ExtractArgsImpl<std::function<R(Args...)>> {
+  using ReturnType = R;
+  template <size_t i>
+  using Arg = typename std::tuple_element<i, std::tuple<Args...>>::type;
+};
+
 template <typename R, typename... Args>
 struct ExtractArgsImpl<R(Args...)> {
   using ReturnType = R;
@@ -120,6 +139,12 @@ BIND_INTERNAL_EXTRACT_CALLABLE_RUN_TYPE_WITH_QUALS(const noexcept);

 #undef BIND_INTERNAL_EXTRACT_CALLABLE_RUN_TYPE_WITH_QUALS

+// 1. Missing specialization to extract signatures from std::function
+template <typename R, typename... Args>
+struct ExtractCallableRunTypeImpl<std::function<R(Args...)>> {
+  using Type = R(Args...);
+};
+
 // Evaluated to the RunType of the given callable type; e.g.
 // `ExtractCallableRunType<decltype([](int, char*) { return 0.1; })>` ->
 //     `double(int, char*)`.
@@ -134,6 +159,13 @@ using ExtractCallableRunType =
 template <typename Functor, typename... BoundArgs>
 struct FunctorTraits;

+// 3. Missing specialization for FunctorTraits to recognize std::function
+template <typename R, typename... Args>
+struct FunctorTraits<std::function<R(Args...)>> {
+  using RunType = R(Args...);
+  static constexpr size_t kNumArgs = sizeof...(Args);
+};
+
 // Provides functor traits for pre-decayed functor types.
 template <typename Functor, typename... BoundArgs>
 struct DecayedFunctorTraits;
diff --git build/config/win/BUILD.gn build/config/win/BUILD.gn
index 4e4ab177d..e28debb68 100644
--- build/config/win/BUILD.gn
+++ build/config/win/BUILD.gn
@@ -222,6 +222,8 @@ config("runtime_library") {
   cflags = []
   cflags_cc = []

+  ldflags = [ "/LIBPATH:" + getenv("VCINSTALLDIR") + "\\Tools\\MSVC\\" + getenv("VCToolsVersion") + "\\lib\\x64" ]
+
   # Defines that set up the CRT.
   defines = [
     "__STD_C",
diff --git build/toolchain/win/tool_wrapper.py build/toolchain/win/tool_wrapper.py
index 47bbfe2a0..3bec9006d 100644
--- build/toolchain/win/tool_wrapper.py
+++ build/toolchain/win/tool_wrapper.py
@@ -140,7 +140,7 @@ class WinTool(object):
     # Read output one line at a time as it shows up to avoid OOM failures when
     # GBs of output is produced.
     for line in link.stdout:
-      line = line.decode('utf8')
+      line = line.decode('cp1252')
       if (not line.startswith('   Creating library ')
           and not line.startswith('Generating code')
           and not line.startswith('Finished generating code')):
